### PR TITLE
DS Prefill :: Add fp8 quantization and compression ops

### LINKED ISSUE #44908: DS Prefill :: FP8 quantization and compression
Per-Token Cast Utilities
DeepEP provides high-performance Python utilities in deep_ep/utils/math.py for manual quantization management, which is useful for pre-processing before calling dispatch.

`per_token_cast_to_fp8`
This function converts a BF16/FP32 tensor to FP8 (E4M3) and extracts scale factors.

Input: x (Tensor), num_bits (default 8).
Mechanism: Calculates the maximum absolute value per token (or per channel block) and computes a scaling factor to map the range to the FP8 representable space (e.g., 448.0 for E4M3).
Output: (x_fp8, scales).

`per_token_cast_back`
Reverses the process, converting FP8 data back to BF16/FP32 using the provided scale factors.

Input: x_fp8, scales.
Mechanism: Performs element-wise multiplication of the dequantized FP8 value by the corresponding scale.

Files:
deep_ep/utils/math.py
tests/elastic/test_ep.py

Src: https://deepwiki.com/deepseek-ai/DeepEP/6.4-fp8-quantization-and-compression

### PR DESCRIPTION
**Pull request by nmauriceTT**
### PR Category

#44908

### Summary

This PR adds two operations:
- `per_token_cast_to_fp8`: take a bfloat16/float32 tensor and quantize to e4m3 + float32 scale tensor.  
- `per_token_cast_back`: take a e4m3 and a float32 scale tensor and computes the dequantized tensor (reverse operation). 

### Notes for reviewers

Operations only support the following configurations:
- It is Blackhole-only.
- Row-major inputs and outputs tensors only
- Tensors height can be any size; but tensor width must be multiple of 128. 
- Scale is float32-only. 
- Core distribution is done on a block-per-block basis (c.f. does not use all cores on small shapes). 

Ops accuracy has been validated with both PCC (> 0.999) and allclose.

#### Implementation details

**per_token_cast_to_fp8** follows the formula in the DeepEP repository:
1. For each segment of 128 consecutive elements (spread across 4 tiles), it computes the absolute value.
2. It compute max absolute value by reduce each tile separately, and then calling `max` LLK between tiles. This puts the largest absolute value on the first column. 
3. To normalize by this value, kernel broadcast the value to a full tile, clamp it, and divide input by this value value. 
4. Normalize value is then quantized to to e4m3 (after utilizing)
5. In parallel, the column of maximum values is append to the tensor of max values (1 column at a time, and this is done in writer kernel). 

As we have 4 tiles, the described is done at "batch" 32. 

Unlike tile layout tensors with have 'fixed' sized pages (typically 1024 elements), pages with row-major tensor take up a full row. If we were to read/write 128 elements at a time, DRAM utilization would be quite inefficient.
To avoid this, data-movement will try to read/write as much data as possible to/from DRAM (minimize number of NoC transaction on same bank); and fill the 'block cb'. 

**per_token_cast_back** also follows the formula:
1. For each segment of 128 consecutive elements (spread. across 4 tiles), it read 1 column of 'scale'
2. Scale column is broadcasted by compute kernel to a full tile and e4m3 input (then converted to float32 in tensix register) is multiplied by this scale.
3. Output dequantized value is written (after untilize). 

Like with `per_token_cast_to_fp8`, DRAM-utilization is increased by a block-based approach. 
 

#### Performance

On Blackhole P100:

##### `per_token_cast_to_fp8`

impl_type | shape | core_count | avg_duration_ms | host_duration_ms | total_duration_ms | overall_bw_gelems_s | Mem Device [GB/s]
-- | -- | -- | -- | -- | -- | -- | --
bf16 | [128, 7168] | 110 | 0.03 | 0.15 | 0.18 | 5.20 | 97.4
bf16 | [640, 7168] | 110 | 0.07 | 0.16 | 0.22 | 20.39 | 208.9
bf16 | [1024, 7168] | 110 | 0.10 | 0.14 | 0.24 | 30.49 | 212.9
bf16 | [3200, 7168] | 110 | 0.29 | 0.14 | 0.43 | 53.08 | 239.0
bf16 | [4096, 7168] | 110 | 0.37 | 0.15 | 0.51 | 57.18 | 242.8
bf16 | [6400, 7168] | 110 | 0.56 | 0.14 | 0.71 | 64.78 | 246.2
fp32 | [128, 7168] | 110 | 0.03 | 0.14 | 0.17 | 5.30 | 136.4
fp32 | [640, 7168] | 110 | 0.11 | 0.16 | 0.27 | 17.22 | 215.4
fp32 | [1024, 7168] | 110 | 0.16 | 0.15 | 0.31 | 23.53 | 226.6
fp32 | [3200, 7168] | 110 | 0.46 | 0.14 | 0.60 | 38.17 | 249.7
fp32 | [4096, 7168] | 110 | 0.58 | 0.14 | 0.72 | 40.83 | 254.4
fp32 | [6400, 7168] | 110 | 0.90 | 0.14 | 1.04 | 44.16 | 256.1

##### `per_token_cast_back `

dtype | Shape | #Cores | Device [ms] | Host [ms] | Total [ms] | Throughput [GElem/s] | Mem Device [GB/s]
-- | -- | -- | -- | -- | -- | -- | --
bf16 | [128, 7168] | 110 | 0.02 | 0.15 | 0.17 | 5.49 | 137.9
bf16 | [640, 7168] | 110 | 0.08 | 0.13 | 0.22 | 21.14 | 168.1
bf16 | [1024, 7168] | 110 | 0.13 | 0.13 | 0.27 | 27.40 | 166.1
bf16 | [3200, 7168] | 110 | 0.41 | 0.14 | 0.55 | 41.84 | 168.4
bf16 | [4096, 7168] | 110 | 0.52 | 0.13 | 0.65 | 44.92 | 170.8
bf16 | [6400, 7168] | 110 | 0.81 | 0.14 | 0.95 | 48.33 | 170.9
fp32 | [128, 7168] | 110 | 0.03 | 0.14 | 0.18 | 5.23 | 147.3
fp32 | [640, 7168] | 110 | 0.14 | 0.16 | 0.31 | 14.95 | 160.2
fp32 | [1024, 7168] | 110 | 0.23 | 0.13 | 0.36 | 20.20 | 160.7
fp32 | [3200, 7168] | 110 | 0.71 | 0.13 | 0.84 | 27.17 | 162.7
fp32 | [4096, 7168] | 110 | 0.90 | 0.14 | 1.04 | 28.37 | 164.5
fp32 | [6400, 7168] | 110 | 1.40 | 0.14 | 1.54 | 29.83 | 164.5

Both `per_token_cast_to_fp8 ` and `per_token_cast_back` seem memory/NoC bound: if they compute bound, the bfloat16 and float32 throughput (in element/s) would be comparable (same compute kernel path), and they would generate different bandwidth. But that's not the case here. 

L1 Nightly fails due to unrelated tests, but running new tests on custom test dispatch give green CI:
- WH: https://github.com/tenstorrent/tt-metal/actions/runs/27020286065
- BH: https://github.com/tenstorrent/tt-metal/actions/runs/27020317843

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:nmaurice/quantization-fp8)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:nmaurice/quantization-fp8)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:nmaurice/quantization-fp8)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:nmaurice/quantization-fp8)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:nmaurice/quantization-fp8)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:nmaurice/quantization-fp8)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:nmaurice/quantization-fp8)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:nmaurice/quantization-fp8)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/common/fp8_quant_common.hpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+inline ttnn::Shape scale_shape_from_input(const ttnn::Shape& input_shape) {
+    const auto rank = input_shape.size();
+    const uint32_t H = static_cast<uint32_t>(input_shape[rank - 1]);
+    TT_FATAL(
+        H % SCALE_GROUP_SIZE == 0,
+        "Per-token cast: hidden dim H={} must be a multiple of SCALE_GROUP_SIZE={}",
+        H,
+        SCALE_GROUP_SIZE);
+    ttnn::SmallVector<uint32_t> dims;
```
[the remark this answers is not in the document]
Comment (nmauriceTT) [reply ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/common/fp8_quant_common.hpp]: Fixed.

**Review by bbradelTT — APPROVED** [review]

**Review by mbezuljTT — APPROVED** [review]

File: tests/ttnn/nightly/unit_tests/operations/experimental/deepseek_prefill/test_deepseek_prefill_per_token_cast.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+    H = 1024
+    M = 128 * 16
+    # In this test, we want to verify that cast_to_fp8 scale is lossless if 128-consecutives values are the same
+    # and if they can be represented exactly in fp8.
+    # This generate values [-M/4, -M/4 + 0.25, ..., M/4 - 0.25] (centered around 0)
+    # And for each value, repeat 128 times horizontally; and 1024 times vertically.
+    STEP = 0.25
+    MIN_VAL = -M / 2 * 0.25
+    x = torch.tensor([(MIN_VAL + i * STEP) for i in range(M // 128)], dtype=torch_dtype).repeat([H, M])
```
[the remark this answers is not in the document]
Comment (nmauriceTT) [reply tests/ttnn/nightly/unit_tests/operations/experimental/deepseek_prefill/test_deepseek_prefill_per_token_cast.py]: Fixed

**Review by dgomezTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/per_token_cast_back/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,56 @@
+include(sources.cmake)
+
+add_library(ttnn_op_experimental_deepseek_prefill_per_token_cast_back ${LIB_TYPE})
+add_library(
+    TTNN::Ops::Experimental::DeepSeekPrefill::PerTokenCastBack
+    ALIAS ttnn_op_experimental_deepseek_prefill_per_token_cast_back
+)
+
+target_precompile_headers(ttnn_op_experimental_deepseek_prefill_per_token_cast_back REUSE_FROM TTNN::PCH)
```
Comment (nsextonTT) [inline ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/per_token_cast_back/CMakeLists.txt]:
```suggestion
target_precompile_headers(ttnn_op_experimental_deepseek_prefill_per_token_cast_back REUSE_FROM TTNN::PCHFull)
```

The standard PCH was split up into PCHLean and PCHFull to reduce compile times. Since you are using device_operation.hpp, the PCHFull variant is the one you want.

**Review by nsextonTT — APPROVED** [review]
Approving to unblock. Please apply comment suggestion.

File: ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/per_token_cast_back/CMakeLists.txt (hunk 3 above)
Comment (nmauriceTT) [reply ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/per_token_cast_back/CMakeLists.txt]: Thanks. 
I've changed `TTNN::PCH` to `TTNN::PCHFull` for  both `per_token_cast_to_fp8` and `per_token_cast_back`.
